d3-bar-chart-race/src/App.js

```diff
@@ -0,0 +1,361 @@
+              )
+        )
+        .call((bar) =>
+          bar
+            .transition(transition)
+            .attr("transform", (d) => `translate(${x(d.value)},${y(d.rank)})`)
+            .call((g) =>
+              g
+                .select("tspan")
+                .tween("text", (d) =>
+                  textTween((prev.get(d) || d).value, d.value)
+                )
+            )
+        ));
+  };
+
+  // draw axis
+  const axis = (svg) => {
+    const g = svg.append("g").attr("transform", `translate(0,${margin.top})`);
+    const axis = d3
+      .axisTop(x)
+      .ticks(width / 150)
+      .tickSizeOuter(0)
+      .tickSizeInner(-barSize * (n + y.padding()));
+
+    return (_, transition) => {
+      g.transition(transition).call(axis);
+      g.select(".tick:first-of-type text").remove();
+      g.selectAll(".tick:not(:first-of-type) line").attr("stroke", "white");
+      g.select(".domain").remove();
+    };
+  };
+
+  // draw date ticker
+  const ticker = (svg) => {
+    if (keyframes.length) {
+      const now = svg
+        .append("text")
+        .style("font", `bold ${barSize}px sans-serif`)
+        .style("font-variant-numeric", "tabular-nums")
+        .attr("text-anchor", "end")
+        .attr("x", width - 30)
+        .attr("y", margin.top + barSize * (n - 0.45))
+        .attr("dy", "0.32em")
+        .text(formatGroupingDate()(keyframes[0][0]));
+
+      return ([date], transition) => {
+        transition.end().then(() => now.text(formatGroupingDate()(date)));
+      };
+    }
+  };
+
+  function* myGenerator() {
+    const svg = d3.select(ref).attr("viewBox", [0, 0, width, height]);
+    svg.selectAll("*").remove();
+
+    const updateBars = bars(svg);
+    const updateAxis = axis(svg);
+    const updateLabels = labels(svg);
+    const updateTicker = ticker(svg);
+
+    for (const keyframe of keyframes) {
+      const transition = svg
+        .transition()
+        .duration(duration)
+        .ease(d3.easeLinear);
+
+      // Extract the top bar’s value.
+      x.domain([0, keyframe[1][0].value]);
+
+      updateAxis(keyframe, transition);
+      updateBars(keyframe, transition);
+      updateLabels(keyframe, transition);
+      updateTicker(keyframe, transition);
+      yield transition.end();
+    }
+  }
+  return myGenerator();
+}
+
+function App() {
+  const config = useConfig();
+  const columns = useElementColumns(config.source);
+  const sigmaData = useElementData(config.source);
+  const [ref, setRef] = useState();
+  const [iter, setIter] = useState(null);
+  const [replay, setReplay] = useState(false);
+  const [firstClick, setFirstClick] = useState(true);
+  const transformedData = useMemo(
+    () => transform(config, columns, sigmaData),
+    [config, columns, sigmaData]
+  );
+
+  useMemo(
+    () => setIter(getChart(transformedData, ref)),
+    [transformedData, ref]
+  );
+
+  if (replay) {
+    setFirstClick(false);
+    setReplay(!replay);
+    setIter(getChart(transformedData, ref));
+  }
```

Comment:
can this happen as part of a callback instead of a conditional block during render? e.g. 
```
const replay = useCallback(() => {
  setReplay(r => !r);
  setFirstClick(false);
}, []);
```

Reply:
Yeah, I have refactored it to make it more concise.